### DISCUSSION TITLE
Decide on a Rust server framework

### DISCUSSION THREAD

**Discussion by mathe-matician** [jazzyclimber/beej-zach-app#3] [General]
If we are pretty much in agreement that we want to use Rust, we should take the next steps in choosing a framework for the web server. 

It looks like there are a few top tools already in use:
- [https://actix.rs/](https://actix.rs/)
- [https://rocket.rs/v0.5-rc/overview/](https://rocket.rs/v0.5-rc/overview/)
- [https://docs.rs/warp/latest/warp/](https://docs.rs/warp/latest/warp/)

I'd be down for the first `Actix` or even `Rocket`. They both seem fairly straight forward and have good documentation.

I suggest these frameworks as well since they are compatible with [Juniper](https://graphql-rust.github.io/), a GraphQL server library for Rust. (Again, the assumption being we are using GraphQL - which I'm down for!)

Just noting here too - although it seems that a lot of people use Juniper, their docs have the disclaimer: `Juniper has not reached 1.0 yet, thus some API instability should be expected.`.

**Comment by jazzyclimber** [comment]
### Rocket
I think that this one has my vote.

Warp, hyper, and tide all seem pretty straight forward but I think that Rocket's syntax resonates with me a bit more, at this point at least. 

### actix
So did a bit of looking round and I am reading a fair amount of negative talk with regards to actix. Mainly it seems like the crew manages it is a bit on the fritz. I guess the head guy once threatened to take all of the repos offline. Which doesnt sound promising haha. Also folks are talking about how bad the docs are ... or at least used to be. 

## GraphQl
It seems like Juniper and Async_graphql are the two libraries at the moment for GraphQl. Def seems like Juniper is the one to go with. And I know they have been around on other languages as well. At least I feel like i've seen the name a few times before. 

I think we could hold off on setting it up -- however, the majority of our final app will be populated almost entirely on API calls. So I def think that having GraphQL set up correctly would really help with limiting server calls. So it's likely worth the initial investment of time to get it rolled up. 

I do worry about the stability issue of juniper -- but folks seem to say that they are using in in production currently and that it's not much of an issue so I think for now it will make sense. And with the amount of time it will likely take us to get anything  significant rolling, they might be at v1. 

Thoughts?

**Comment by mathe-matician** [comment]
Yeah, if that's true about Actix, let's just go with Rocket.

I'm not too familiar on GraphQL's place in the stack, in relation to when we should set it up that is. Could you go into this more? A high level explanation is fine for now. Is it that you create your API and then pass your calls through GraphQL? What is the typical setup?

Can you also setup [Github Discussions](https://docs.github.com/en/discussions/quickstart) - I feel like we will have a few of these types of threads that seem more appropriate for Discussions rather than creating issues. We can then move this to a thread in there.

**Reply by jazzyclimber** [reply]
So I've pretty limited experience in configuring a graphQL server. I did it once for a small project that I ditched. But GraphQL is the endpoint. It's a single graphQL endpoint that accesses your data and is configured so that it spits out typed data in a very predictable way. So a front end dev knows exactly what they are getting and what the data will be coming out. So they don't have to write much data validation. 

It also is helpful because you can grab all of the data you need in a single call as oppose to needing to ping multiple endpoint to grab different data points. 

[This article](https://dev.to/pratham10/does-graphql-replace-rest-api--50fm) sums up some key points but it's not really technical but gives a gist of what it's all about

**Reply by mathe-matician** [reply]
I actually read that exact article the other day - I didn't look at the comments, though. Probably worth keeping in mind/investigating these two when we start building things out.

> GraphQL has a lot of drawbacks in comparison to REST ranging from bloated payload to security and scalability concerns. But if one uses GraphQL on the edge of infrastructure, e.g. to connect RESTful microservices, it truly shines. Hence those two are not rivals, they compliment one another quite nicely

>... Authorization (not Authentication) and DoS prevention, are a nightmare in GraphQL.

**Reply by jazzyclimber** [reply]
Ahhh yea I didnt look through the comments either. 

I can chat with Gary about it this week too (my bossman). I think that they are using it on this booth-central/boothy product they made. Which they also built from the ground up. Ill see what his thoughts are.

**Reply by jazzyclimber** [reply]
Spoke with Gary's brother yesterday and they are using it on their app. 

They really like the ease at which they are able to make calls and grab data -- especially knowing exactly what they are going to get back. 

His main concern with it was that -- on Laravel (their backend) -- there is only one library that supports graphql so when they added  it to their app there wasnt much documentation on it. 

I feel like we may be in a similar boat with RUST in general. I just saw that Rocket is actually still not V1. 

He also brought up some good points on using newer frameworks and languages. Primary concern being that they may be good for smaller apps and work well within that environment but it's not uncommon to start building and then realize that the thing you thought was going to be dope -- is actually difficult to implement and use on a larger ecosystem/app. 

So maybe something to keep in mind. 

------

I think what this means for us is that if we are going to use a RUST backend the place to start might be spending the next week or so just learning the language a bit. Then planning out some of the high level parts of the app and what the server/backend might need to do all around. We can then lay out that plumbing and if we see any concerns we can pivot before we get too vested in on path.

Same goes for the frontend. I'd like to use svelte just to learn and play with it. But it might make more sense to use react or vue. Im much more comfy with vue. And I personally like it's syntax better. But if svelte is an option I think it would be really fun to build it in there.

**Reply by mathe-matician** [reply]
That makes sense for newer languages and such, it's definitely a given. From brief searches Rust's adoption seem fairly decent so far - that doesn't necessarily solve the lack of tools created by the community though. It does seem like a neat language from what I've learned so far, and generally a fun learning opportunity. We do have plenty of other options, though.

I agree - they are definitely things to keep in mind during planning and development. Ultimately, though, I'd much rather use Rust and Svelte over the other options.

Those steps forward seem reasonable - we can always talk more on Saturday if we decide to meet up, or just start hashing stuff out in another discussion thread.
